src/parser_yang.c

```diff
@@ -4516,6 +4519,64 @@ yang_check_deviate(struct lys_module *module, struct unres_schema *unres, struct
 
         /* ... and replace it with the value specified in deviation */
         dev_target->flags |= deviate->flags & LYS_CONFIG_MASK;
+
+        /* "config" is explicitely set in the node */
+        dev_target->flags |= LYS_CONFIG_SET;
+
+        /* "config" must be set exclusively to "yes" or "no" */
+        assert(!((dev_target->flags & LYS_CONFIG_W) && (dev_target->flags & LYS_CONFIG_R)));
+
+        /* from rfc7950#page-86: All key leafs in a list MUST have the same value
+         * for their "config" as the list itself.
+         */
+        if ((node_list = lys_is_key((const struct lys_node_leaf *)dev_target, NULL))) {
+            if ((node_list->flags & LYS_CONFIG_MASK) != (dev_target->flags & LYS_CONFIG_MASK)) {
+                LOGVAL(module->ctx, LYE_INSTMT, LY_VLOG_NONE, NULL, "config");
+                LOGVAL(module->ctx, LYE_SPEC, LY_VLOG_NONE, NULL, "The deviate causes a violation of \"config\" in key leaf.");
+                goto error;
+            }
+        }
+
+        /* from rfc7950#page-135: If the parent node is a case node,
+         * the value is the same as the case node's parent choice node.
+         */
+        if (dev_target->nodetype == LYS_CASE) {
+            if ((parent = lys_parent(dev_target)) && (parent->nodetype == LYS_CHOICE)) {
+                if ((parent->flags & LYS_CONFIG_MASK) != (dev_target->flags & LYS_CONFIG_MASK)) {
+                    LOGVAL(module->ctx, LYE_INSTMT, LY_VLOG_NONE, NULL, "config");
+                    LOGVAL(module->ctx, LYE_SPEC, LY_VLOG_NONE, NULL, "The deviate causes a violation of \"config\" in case node.");
+                    goto error;
+                }
+            }
+        }
+
+        /* from rfc7950#page-135: If a node has "config" set to "false",
+         * no node underneath it can have "config" set to "true".
+         */
+        if (dev_target->flags & LYS_CONFIG_R) {
+            struct lys_node *next, *elem;
+            LY_TREE_DFS_BEGIN(dev_target->child, next, elem) {
+                if (elem->flags & LYS_CONFIG_W) {
+                    if (elem->flags & LYS_CONFIG_SET) {
+                        LOGVAL(module->ctx, LYE_INSTMT, LY_VLOG_NONE, NULL, "config");
+                        LOGVAL(module->ctx, LYE_SPEC, LY_VLOG_NONE, NULL, "The deviate causes a violation of \"config\" in child node.");
+                        goto error;
+                    } else {
+                        /* remove current config value of the elem including LYS_CONFIG_SET */
+                        elem->flags &= ~(LYS_CONFIG_MASK | LYS_CONFIG_SET) ;
```

Comment:
Please pay attention here if the **LYS_CONFIG_SET** bit should really be cleared. The point is that if the **config** was explicitly set to **yes** in the node, and at the same time the inheritance from the ancestor was projected, then the **config** is switched to **no**, so such a **config** does not have the value set explicitly, but rather implicitly. Therefore, the flag is deleted.

Reply:
I think it is a good habit to capture these possible gotchas as comments in the code rather that explanatory comments in the code review. The latter tends to get lost quicker.